Kudu.Services/Deployment/PushDeploymentController.cs

```diff
@@ -354,13 +517,70 @@ bool isFunctionJson(string fullName)
         {
-            var content = await DeploymentHelper.GetZipContentFromURL(zipDeploymentInfo, tracer);
+            var artifactDeploymentInfo = (ArtifactDeploymentInfo) deploymentInfo;
+
+            // This is the path where the artifact being deployed is staged, before it is copied to the final target location
+            var artifactDirectoryStagingPath = repository.RepositoryPath;
+
+            var targetInfo = FileSystemHelpers.DirectoryInfoFromDirectoryName(artifactDirectoryStagingPath);
+            if (targetInfo.Exists)
+            {
+                // If tempDirPath already exists, rename it so we can delete it later 
+                var moveTarget = Path.Combine(targetInfo.Parent.FullName, Path.GetRandomFileName());
+                using (tracer.Step(string.Format("Renaming ({0}) to ({1})", targetInfo.FullName, moveTarget)))
+                {
+                    targetInfo.MoveTo(moveTarget);
+                }
+            }
+
+            // Create artifact staging directory before later use 
+            Directory.CreateDirectory(artifactDirectoryStagingPath);
+            var artifactFileStagingPath = Path.Combine(artifactDirectoryStagingPath, deploymentInfo.TargetFileName);
+
+            // If RemoteUrl is non-null, it means the content needs to be downloaded from the Url source to the staging location
+            // Else, it had been downloaded already so we just move it to the staging location
+            if (!string.IsNullOrWhiteSpace(artifactDeploymentInfo.RemoteURL))
+            {
+                using (tracer.Step("Saving request content to {0}", artifactFileStagingPath))
+                {
+                    var content = await DeploymentHelper.GetArtifactContentFromURL(artifactDeploymentInfo, tracer);
+                    var copyTask = content.CopyToAsync(artifactFileStagingPath, tracer);
+
+                    // Deletes all files and directories except for artifactFileStagingPath, artifactDirectoryStagingPath, and tracer
+                    var cleanTask = Task.Run(() => DeleteFilesAndDirsExcept(artifactFileStagingPath, artifactDirectoryStagingPath, tracer));
+
+                    // Lets the copy and cleanup tasks to rin n parallel and wait for them to finish 
+                    await Task.WhenAll(copyTask, cleanTask);
+                }
+            }
+            else
+            {
+                var srcInfo = FileSystemHelpers.DirectoryInfoFromDirectoryName(deploymentInfo.RepositoryUrl);
+                using (tracer.Step(string.Format("Moving ({0}) to ({1})", targetInfo.FullName, artifactFileStagingPath)))
+                {
+                    srcInfo.MoveTo(artifactFileStagingPath);
+                }
+
+                // Deletes all files and directories except for artifactFileStagingPath, artifactDirectoryStagingPath, and tracer
+                DeleteFilesAndDirsExcept(artifactFileStagingPath, artifactDirectoryStagingPath, tracer);
```

Comment:
tracer is used for tracing, it is not a file path
// Deletes all files and directories except artifactFileStagingPath and artifactDirectoryStagingPath